airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.testutils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.JdbcDatabaseContainer;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.utility.DockerImageName;
+
+/**
+ * ContainerFactory is the companion interface to {@link TestDatabase} for providing it with
+ * suitable testcontainer instances.
+ */
+public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+
+  /**
+   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
+   * the testcontainer type.
+   */
+  C createNewContainer(DockerImageName imageName);
+
+  /**
+   * Returns the class object of the testcontainer.
+   */
+  Class<?> getContainerClass();
+
+  /**
+   * Returns a shared instance of the testcontainer.
+   */
+  default C shared(String imageName, String... methods) {
+    final String mapKey = Stream.concat(
+        Stream.of(imageName, this.getClass().getCanonicalName()),
+        Stream.of(methods))
+        .collect(Collectors.joining("+"));
+    return Singleton.getOrCreate(mapKey, this);
+  }
+
+  /**
+   * This class is exclusively used by {@link #shared(String, String...)}. It wraps a specific shared
+   * testcontainer instance, which is created exactly once.
+   */
+  class Singleton<C extends JdbcDatabaseContainer<?>> {
```

Comment:
private static

Reply:
I'd like to, but I can't, because this inner class' outer class is an interface. This means the `static` is implied and the `private` is not possible.